crates/loro-internal/src/encoding/arena.rs

```diff
@@ -0,0 +1,463 @@
+use std::{borrow::Cow, ops::Deref};
+
+use crate::InternalString;
+use fxhash::FxHashSet;
+use itertools::Itertools;
+use loro_common::{ContainerID, ContainerType, Counter, LoroError, LoroResult, PeerID};
+use serde::{Deserialize, Serialize};
+use serde_columnar::{columnar, ColumnarError};
+
+use super::encode_reordered::{PeerIdx, ValueRegister, MAX_DECODED_SIZE};
+
+pub(super) fn encode_arena(
+    peer_ids_arena: Vec<u64>,
+    containers: ContainerArena,
+    keys: Vec<InternalString>,
+    deps: DepsArena,
+    position_arena: PositionArena,
+    tree_id_arena: TreeIDArena,
+    state_blob_arena: &[u8],
+) -> Vec<u8> {
+    let peer_ids = PeerIdArena {
+        peer_ids: peer_ids_arena,
+    };
+
+    let key_arena = KeyArena { keys };
+    let encoded = EncodedArenas {
+        peer_id_arena: &peer_ids.encode(),
+        container_arena: &containers.encode(),
+        key_arena: &key_arena.encode(),
+        deps_arena: &deps.encode(),
+        position_arena: &position_arena.encode(),
+        tree_id_arena: &tree_id_arena.encode(),
+        state_blob_arena,
+    };
+
+    encoded.encode_arenas()
+}
+
+pub struct EncodedRegisters<'a> {
+    pub(super) peer: ValueRegister<PeerID>,
+    pub(super) key: ValueRegister<InternalString>,
+    pub(super) container: ValueRegister<ContainerID>,
+    pub(super) tree_id: ValueRegister<EncodedTreeID>,
+    pub(super) position: either::Either<FxHashSet<&'a [u8]>, ValueRegister<&'a [u8]>>,
+}
+
+impl<'a> EncodedRegisters<'a> {
+    pub(crate) fn convert_position(&mut self) {
+        let position_register =
+            std::mem::replace(&mut self.position, either::Left(Default::default()))
+                .left()
+                .unwrap();
+        let positions = position_register.into_iter().sorted_unstable().collect();
+        let position_register = ValueRegister::from_existing(positions);
+        self.position = either::Right(position_register);
+    }
+}
+
+pub struct DecodedArenas<'a> {
+    pub(super) peer_ids: PeerIdArena,
+    pub(super) containers: ContainerArena,
+    pub(super) keys: KeyArena,
+    pub deps: Box<dyn Iterator<Item = Result<EncodedDep, ColumnarError>> + 'a>,
+    pub(super) positions: PositionArena<'a>,
+    pub(super) tree_ids: TreeIDArena,
+    pub state_blob_arena: &'a [u8],
+}
+
+pub fn decode_arena(bytes: &[u8]) -> LoroResult<DecodedArenas> {
+    let arenas = EncodedArenas::decode_arenas(bytes)?;
+    Ok(DecodedArenas {
+        peer_ids: PeerIdArena::decode(arenas.peer_id_arena)?,
+        containers: ContainerArena::decode(arenas.container_arena)?,
+        keys: KeyArena::decode(arenas.key_arena)?,
+        deps: Box::new(DepsArena::decode_iter(arenas.deps_arena)?),
+        positions: PositionArena::decode(arenas.position_arena)?,
+        tree_ids: TreeIDArena::decode(arenas.tree_id_arena)?,
+        state_blob_arena: arenas.state_blob_arena,
+    })
+}
+
+struct EncodedArenas<'a> {
```

Comment:
Is this extendible? Can I add more fields to it?

Reply:
Yes, the part of `arena` is at the end of encoded bytes, it is extendible.
You can add new fields at the end in order, the newer (unknown) part will be ignored when decoded by older version.